### PR TITLE
WIP feat: use firestore instead of env config for kms key storing

### PR DESCRIPTION
**Pull request by capri-xiyue**
use firestore instead of env config for kms key storing

### CODE REVIEW

File: pkg/testutil/firestore_mock.go [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+	// If set, all calls return this error.
+	Error error
+
+	// responses to return if err == nil
+	Resps []proto.Message
+}
+
+func (s *MockFirestoreServer) GetDocument(ctx context.Context, req *firestorepb.GetDocumentRequest) (*firestorepb.Document, error) {
+	md, _ := metadata.FromIncomingContext(ctx)
+	if xg := md["x-goog-api-client"]; len(xg) == 0 || !strings.Contains(xg[0], "gl-go/") {
```
Comment (capri-xiyue) [inline pkg/testutil/firestore_mock.go:61]: the code borrowed from here https://github.com/googleapis/google-cloud-go/blob/6d1e9fe399d0f9adf1a80107be1b27c725b02571/firestore/apiv1/mock_test.go#L63.
I tried to investigate why we need such metadata validation here, but can't find the clues.

File: cmd/cert-rotation/main.go [hunk 2]
```diff
@@ -88,8 +95,13 @@ func realMain(ctx context.Context) error {
 		return fmt.Errorf("failed to load config: %w", err)
 	}
 
+	fsClient, err := firestore.NewClient(ctx, config.ProjectID)
+	if err != nil {
+		return fmt.Errorf("failed to setup FireSore client: %w", err)
```
Comment (raserva) [inline cmd/cert-rotation/main.go:100]:
```suggestion
		return fmt.Errorf("failed to setup FireStore client: %w", err)
```

File: cmd/justification/main.go [hunk 3]
```diff
@@ -64,12 +65,17 @@ func realMain(ctx context.Context) error {
 		return fmt.Errorf("failed to setup kms client: %w", err)
 	}
 
+	fsClient, err := firestore.NewClient(ctx, cfg.ProjectID)
+	if err != nil {
+		return fmt.Errorf("failed to setup FireSore client: %w", err)
```
Comment (raserva) [inline cmd/justification/main.go:70]:
```suggestion
		return fmt.Errorf("failed to setup FireStore client: %w", err)
```

File: cmd/public-key/main.go [hunk 4]
```diff
@@ -62,10 +63,15 @@ func realMain(ctx context.Context) error {
 		return fmt.Errorf("failed to load config: %w", err)
 	}
 
+	fsClient, err := firestore.NewClient(ctx, config.ProjectID)
+	if err != nil {
+		return fmt.Errorf("failed to setup FireSore client: %w", err)
```
Comment (raserva) [inline cmd/public-key/main.go:68]:
```suggestion
		return fmt.Errorf("failed to setup FireStore client: %w", err)
```

File: pkg/config/crypto_config.go [hunk 5]
```diff
@@ -45,11 +45,8 @@ type CryptoConfig struct {
 	// DisabledPeriod is a time between when the key is disabled, and when we delete the key.
 	DisabledPeriod time.Duration `yaml:"disabled_period,omitempty" env:"DISABLED_PERIOD,overwrite"`
 
-	// TODO: This is intended to be temporary, and will eventually be retrieved from a persistent external datastore
-	// https://github.com/abcxyz/jvs/issues/17
-	// KeyName format: `projects/*/locations/*/keyRings/*/cryptoKeys/*`
-	// https://pkg.go.dev/google.golang.org/genproto/googleapis/cloud/kms/v1#CryptoKey
-	KeyNames []string `yaml:"key_names,omitempty" env:"KEY_NAMES,overwrite"`
+	// ProjectID is the ID of GCP project where the Firestore documents with the KMS key locates
```
Comment (raserva) [inline pkg/config/crypto_config.go:48]:
```suggestion
	// ProjectID is the ID of GCP project where the Firestore document with the KMS key is located
```

File: pkg/testutil/firestore_mock.go (hunk 1 above)
Comment (raserva) [reply pkg/testutil/firestore_mock.go:61]: Seems to be ensuring that the firestore client is identifying itself. Probably unnecessary for us. 

https://cloud.google.com/apis/docs/system-parameters

**Review by raserva — COMMENTED** [review]
I don't see anything in this change that works with primary key information, is that intended to be done in a future PR?

File: cmd/cert-rotation/main.go [hunk 6]
```diff
@@ -38,12 +40,17 @@ type server struct {
...
+	ctx := r.Context()
+	logger := logging.FromContext(ctx)
 	logger.Info("received request", zap.Any("url", r.URL))
 
 	var errs error
-	// TODO: load keys from DB instead. https://github.com/abcxyz/jvs/issues/17
-	for _, key := range s.handler.CryptoConfig.KeyNames {
+	kmsConfig, err := fsutil.GetKMSConfig(ctx, s.handler.FsClient, fsutil.Collection, fsutil.CertRotationConfigDoc)
+	if err != nil {
+		logger.Error("ran into errors while getting kms config", zap.Error(err))
```
[the remark this answers is not in the document]
Comment (capri-xiyue) [reply cmd/cert-rotation/main.go:50]: I was just following the code style of this method https://github.com/abcxyz/jvs/blob/5cc3eb83dc6b7e4af19a40c2c20456d2912770f5/cmd/cert-rotation/main.go#L54. In addition, this `ServeHTTP` method doesn't return error

File: cmd/cert-rotation/main.go (hunk 2 above)
Comment (sethvargo) [reply cmd/cert-rotation/main.go:100]: It's actually "Firestore" 😄

File: cmd/justification/main.go [hunk 7]
```diff
@@ -64,12 +65,17 @@ func realMain(ctx context.Context) error {
 		return fmt.Errorf("failed to setup kms client: %w", err)
 	}
 
+	fsClient, err := firestore.NewClient(ctx, cfg.ProjectID)
```
Comment (sethvargo) [inline cmd/justification/main.go:68]: I think we need to close these clients when the server shuts down. Applies to KMS too.

File: pkg/config/crypto_config.go [hunk 8]
```diff
@@ -45,11 +45,8 @@ type CryptoConfig struct {
...
+	ProjectID string `yaml:"project_id,omitempty" env:"PROJECT_ID,overwrite"`
```
Comment (sethvargo) [inline pkg/config/crypto_config.go:49]: This should probably be `FirebaseProjectID`. As-is, we're limiting the ability to have other services in different projects with a generic name.

File: pkg/config/public_key_config.go [hunk 9]
```diff
@@ -19,17 +19,15 @@ import (
...
 
 // PublicKeyConfig is the config used for public key hosting.
 type PublicKeyConfig struct {
-	// TODO: This is intended to be temporary, and will eventually be retrieved from a persistent external datastore
-	// https://github.com/abcxyz/jvs/issues/17
-	// KeyName format: `projects/*/locations/*/keyRings/*/cryptoKeys/*`
-	// https://pkg.go.dev/google.golang.org/genproto/googleapis/cloud/kms/v1#PublicKeyKey
-	KeyNames     []string      `yaml:"key_names,omitempty" env:"KEY_NAMES,overwrite"`
+	// ProjectID is the ID of GCP project where the Firestore documents with the KMS key locates
+	ProjectID    string        `yaml:"project_ID,omitempty" env:"PROJECT_ID,overwrite"`
```
Comment (sethvargo) [inline pkg/config/public_key_config.go:30]: Same note here: this should be `FirebaseProjectID` or similar.

File: pkg/firestore/util.go [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+package firestore
+
+import (
+	"context"
+	"fmt"
+
+	"cloud.google.com/go/firestore"
+)
+
+// [START firestore_data_custom_type_definition]
```
Comment (sethvargo) [inline pkg/firestore/util.go:24]: This comment is copied from the code sample - we don't need it.

File: pkg/firestore/util.go [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+func GetKMSConfig(ctx context.Context, fsClient *firestore.Client, collectionPath, docID string) (*KMSConfig, error) {
+	kSnap, err := fsClient.Collection(collectionPath).Doc(docID).Get(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read from Collection %s Doc %s: %w", collectionPath, docID, err)
+	}
+	var kmsConfig KMSConfig
+	err = kSnap.DataTo(&kmsConfig)
+	if err != nil {
+		return nil, fmt.Errorf("failed to use firestore document's fields to populate struct: %w", err)
+	}
```
Comment (sethvargo) [inline pkg/firestore/util.go:49]:
```suggestion
	if err := kSnap.DataTo(&kmsConfig); err != nil {
		return nil, fmt.Errorf("failed to use firestore document's fields to populate struct: %w", err)
	}
```

File: pkg/firestore/util.go [hunk 12]
```diff
@@ -0,0 +1,51 @@
+// Copyright 2022 Google LLC
```
Comment (sethvargo) [inline pkg/firestore/util.go:1]: Naming things "util" is generally frowned upon in Go.

File: pkg/justification/processor.go [hunk 13]
```diff
@@ -98,7 +102,14 @@ func (p *Processor) CreateToken(ctx context.Context, request *jvspb.CreateJustif
 }
 
 func (p *Processor) getLatestSigner(ctx context.Context) (*signerWithID, error) {
-	ver, err := jvscrypto.GetLatestKeyVersion(ctx, p.kms, p.config.KeyName)
```
Comment (sethvargo) [inline pkg/justification/processor.go:101]: Can we update the comment for `getLatestSigner` to note that it's an expensive operation that communicates with the upstream database?

File: pkg/firestore/util.go [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+	PublicKeyConfigDoc     = "PublicKeyConfig"
+	CertRotationConfigDoc  = "CertRotationConfig"
+)
+
+type KMSConfig struct {
+	// KeyNames format: `[projects/*/locations/*/keyRings/*/cryptoKeys/*]`
+	KeyNames []string `firestore:"key_names,omitempty"`
+}
+
+// [END firestore_data_custom_type_definition]
```
Comment (sethvargo) [inline pkg/firestore/util.go:38]: Same note, comment is not needed

File: pkg/firestore/util.go [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+
+func GetKMSConfig(ctx context.Context, fsClient *firestore.Client, collectionPath, docID string) (*KMSConfig, error) {
```
Comment (sethvargo) [inline pkg/firestore/util.go:40]: What is our strategy for getting this initial data into firestore?

Comment (sethvargo) [inline pkg/firestore/util.go:40]: The collection path and document are constants. Why are they also parameters? If the goal is to make them easier to use in tests, I suggest creating a separate function instead for testing.

File: pkg/jvscrypto/key_hosting.go [hunk 16]
```diff
@@ -19,16 +19,19 @@ import (
...
 	"github.com/abcxyz/jvs/pkg/config"
+	fsutil "github.com/abcxyz/jvs/pkg/firestore"
 	"github.com/abcxyz/pkg/cache"
 	"github.com/abcxyz/pkg/logging"
 )
 
 // KeyServer provides all valid and active public keys in a JWKS format.
 type KeyServer struct {
 	KMSClient       *kms.KeyManagementClient
+	FsClient        *firestore.Client
```
Comment (sethvargo) [inline pkg/jvscrypto/key_hosting.go:34]: This should be `FirestoreClient`

**Review by sethvargo — COMMENTED** [review]
Open question: should we default the value of project ID to looking it up in the environment?

File: pkg/config/public_key_config.go (hunk 9 above)
Comment (yolocs) [reply pkg/config/public_key_config.go:30]: +1

File: pkg/justification/processor.go [hunk 17]
```diff
@@ -40,6 +42,7 @@ import (
 type Processor struct {
 	jvspb.UnimplementedJVSServiceServer
 	kms         *kms.KeyManagementClient
+	fsClient    *firestore.Client
```
Comment (yolocs) [inline pkg/justification/processor.go:45]: We should abstract the implementation away for better encapsulation. So instead of passing in a firestore client, I suggest passing in a `RemoteConfig` (to be defined) interface. The actual implementation will be firestore based.

File: pkg/firestore/util.go [hunk 18]
```diff
@@ -0,0 +1,51 @@
...
+// [START firestore_data_custom_type_definition]
+
+const (
+	Collection             = "JVS"
+	JustificationConfigDoc = "JustificationConfig"
+	PublicKeyConfigDoc     = "PublicKeyConfig"
+	CertRotationConfigDoc  = "CertRotationConfig"
+)
+
+type KMSConfig struct {
```
Comment (yolocs) [inline pkg/firestore/util.go:33]: I think this type is orthogonal to firestore. Thus we should define it in `pkg/config` package.

File: pkg/firestore/util.go (hunk 15 above)
Comment (yolocs) [inline pkg/firestore/util.go:40]: As I mentioned in the other comment, I suggest refactor this to be `RemoteConfig` interface and firestore implementation.

File: pkg/justification/processor_test.go [hunk 19]
```diff
@@ -105,6 +110,14 @@ func TestCreateToken(t *testing.T) {
 		t.Fatal(err)
 	}
 
+	mockFSClient, mockFSServer, mockFsCleanupFunc, err := testutil.NewMockFS(fakeProjectID)
```
Comment (yolocs) [inline pkg/justification/processor_test.go:113]: If we do a good job with encapsulating the `RemoteConfig`, then it seems we wouldn't need implement the whole fake firestore.

File: pkg/jvscrypto/key_hosting.go (hunk 16 above)
Comment (yolocs) [reply pkg/jvscrypto/key_hosting.go:34]: +1. `FsClient` makes me think of filesystem related stuff

**Review by yolocs — CHANGES_REQUESTED** [review]
In the future, try also break up the PR. E.g. the firestore config client, the fake could be its own PR.

**Comment by capri-xiyue** [comment]
> I don't see anything in this change that works with primary key information, is that intended to be done in a future PR?

Do you. mean for [SetPrimary](https://github.com/abcxyz/jvs/blob/5cc3eb83dc6b7e4af19a40c2c20456d2912770f5/pkg/jvscrypto/kmsutil.go#L198) and [getPrimary](https://github.com/abcxyz/jvs/blob/5cc3eb83dc6b7e4af19a40c2c20456d2912770f5/pkg/jvscrypto/kmsutil.go#L184), we want to store the primary key info in the firestore instead of relying on the KMS key labels, right?

**Comment by raserva** [comment]
> > I don't see anything in this change that works with primary key information, is that intended to be done in a future PR?
> 
> Do you mean for [SetPrimary](https://github.com/abcxyz/jvs/blob/5cc3eb83dc6b7e4af19a40c2c20456d2912770f5/pkg/jvscrypto/kmsutil.go#L198) and [getPrimary](https://github.com/abcxyz/jvs/blob/5cc3eb83dc6b7e4af19a40c2c20456d2912770f5/pkg/jvscrypto/kmsutil.go#L184), we want to store the primary key info in the firestore instead of relying on the KMS key labels, right?

Correct. Right now we're storing primary information on key labels, but it should instead be stored in firebase.

**Comment by capri-xiyue** [comment]
> > > I don't see anything in this change that works with primary key information, is that intended to be done in a future PR?
> > 
> > 
> > Do you mean for [SetPrimary](https://github.com/abcxyz/jvs/blob/5cc3eb83dc6b7e4af19a40c2c20456d2912770f5/pkg/jvscrypto/kmsutil.go#L198) and [getPrimary](https://github.com/abcxyz/jvs/blob/5cc3eb83dc6b7e4af19a40c2c20456d2912770f5/pkg/jvscrypto/kmsutil.go#L184), we want to store the primary key info in the firestore instead of relying on the KMS key labels, right?
> 
> Correct. Right now we're storing primary information on key labels, but it should instead be stored in firebase. That can be done as a follow-up if you want.
> 
> But one thing that will be important to think about even in this PR is that we cannot load the config once and be done, instead we need to make sure we have any updates from firebase when we are signing tokens to ensure that we're using the most recent primary

Looks like firestore supports listen to multiple documents(https://firebase.google.com/docs/firestore/query-data/listen#listen_to_multiple_documents_in_a_collection) But in our case, we need long-running listen. Looks like we may have to write some logic to achieve it.

**Comment by yolocs** [comment]
Found https://github.com/spf13/viper#firestore. Can we use that? We already use viper for CLI config.

**Comment by capri-xiyue** [comment]
closed because I created a new PR https://github.com/abcxyz/jvs/pull/91
